File: py/svg.py

import random

DPI = 72
CURVE_TIGHTNESS = 0 # 0 = Catmull-Rom splines, 1 = straight lines

def write_header(f, width, height):
    f.write("""<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="%g" height="%g" viewBox="0 0 %g %g" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
""" % (width, height, width, height))

def draw_line(f, x1, y1, x2, y2):
    f.write("""<line x1="%g" y1="%g" x2="%g" y2="%g" stroke="blue" stroke-width="1"/>
""" %
            (x1, y1, x2, y2))

def draw_rect(f, x1, y1, x2, y2):
    f.write("""<rect x="%g" y="%g" width="%g" height="%g" stroke="blue" stroke-width="1" fill="none"/>
""" %
            (x1, y1, x2 - x1, y2 - y1))

def draw_circle(f, x, y, r):
    f.write("""<circle cx="%g" cy="%g" r="%g" stroke="black" stroke-width="1" fill="none"/>
""" % (x, y, r))

def draw_spline(f, v):
    s = 1 - CURVE_TIGHTNESS
    for i in range(1, len(v) - 2):
        b0 = v[i]
        b1 = (
          v[i][0] + (s * v[i + 1][0] - s * v[i - 1][0]) / 6,
          v[i][1] + (s * v[i + 1][1] - s * v[i - 1][1]) / 6
        )
        b2 = (
          v[i + 1][0] + (s * v[i][0] - s * v[i + 2][0]) / 6,
          v[i + 1][1] + (s * v[i][1] - s * v[i + 2][1]) / 6
        )
        b3 = v[i + 1]
        #draw_line(f, b0[0], b0[1], b1[0], b1[1])
        #draw_line(f, b1[0], b1[1], b2[0], b2[1])
        #draw_line(f, b2[0], b2[1], b3[0], b3[1])
        f.write("""<path d="M %g %g C %g %g, %g %g, %g %g" stroke="black" stroke-width="1" fill="none"/>
""" %
                (b0[0], b0[1], b1[0], b1[1], b2[0], b2[1], b3[0], b3[1]))

def write_footer(f):
    f.write("""</svg>
""")

def lerp(a, b, t):
    return a + (b - a)*t

def duplicate_ends(v):
    return [v[0]] + v + [v[-1]]

def random_in_range(a, b):
    return lerp(a, b, random.random())

File: ts/Svg.ts
import { Circle } from "./Circle.ts";
import { Line } from "./Line.ts";
import { Vector } from "./Vector.ts";

const CURVE_TIGHTNESS = 0; // 0 = Catmull-Rom splines, 1 = straight lines
export const DPI = 96;

/**
 * A class to generate SVG files. All coordinates are in points.
 */
export class Svg {
    private readonly size: Vector;
    private readonly parts: string[] = [];
    private inverted = false;

    /**
     * Construct an SVG page with the given size.
     */
    constructor(size: Vector) {
        this.size = size;

        this.parts.push(`<?xml version="1.0" encoding="UTF-8" standalone="no"?>`);
        this.parts.push(`<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">`);
        this.parts.push(`<svg width="${this.size.x}" height="${this.size.y}" viewBox="0 0 ${this.size.x} ${this.size.y}" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">`);
    }

    public invert() {
        this.parts.push(`<rect x="0" y="0" width="${this.size.x}" height="${this.size.y}" fill="black"/>`);
        this.inverted = true;
    }

    /**
     * Draw a line.
     */
    public drawLine(line: Line, color?: string) {
        if (color === undefined) {
            color = this.inverted ? "white" : "black";
        }
        this.parts.push(`<line x1="${line.p1.x}" y1="${line.p1.y}" x2="${line.p2.x}" y2="${line.p2.y}" stroke="${color}" stroke-width="1"/>`);
    }

    /**
     * Draw a rectangle outline from p1 to p2.
     */
    public drawRect(p1: Vector, p2: Vector) {
        const size = p2.minus(p1);
        this.parts.push(`<rect x="${p1.x}" y="${p1.y}" width="${size.x}" height="${size.y}" stroke="blue" stroke-width="1" fill="none"/>`);
    }

    /**
     * Draw a circle outline centered at "c" with radius "r".
     */
    public drawCircle(c: Vector | Circle, r?: number) {
        if (c instanceof Circle) {
            if (r !== undefined) {
                throw new Error("when circle, can't specify radius");
            }
            r = c.r;
            c = c.c;
        }
        this.parts.push(`<circle cx="${c.x}" cy="${c.y}" r="${r}" stroke="black" stroke-width="1" fill="none"/>`);
    }

    /**
     * Draw an outlined hollow SVG path.
     */
    public drawPath(path: string) {
        const color = this.inverted ? "white" : "black";
        this.parts.push(`<path d="${path}" stroke="${color}" stroke-width="1" fill="none"/>`);
    }

    /**
     * Draw a Catmull-Rom spline. Does not draw the first and last point, so if you want those, duplicate
     * them.
     */
    public drawSpline(v: Vector[]) {
        const s = 1 - CURVE_TIGHTNESS;

        for (let i = 1; i < v.length - 2; i++) {
            const b0 = v[i];
            const b1 = v[i].plus(v[i + 1].minus(v[i - 1]).times(s/6));
            const b2 = v[i + 1].plus(v[i].minus(v[i + 2]).times(s/6));
            const b3 = v[i + 1];
            this.parts.push(`<path d="M ${b0.x} ${b0.y} C ${b1.x} ${b1.y}, ${b2.x} ${b2.y}, ${b3.x} ${b3.y}" stroke="black" stroke-width="1" fill="none"/>`);
        }
    }

    /**
     * Make a single string of the entire SVG file.
     */
    private makeSvgString(): string {
        return this.parts.join("\n") + "\n";
    }

    /**
     * Closes the SVG file and writes it to the given pathname. Do not call multiple times or
     * the file will be closed too many times.
     */
    public async save(pathname: string) {
        this.parts.push(`</svg>`);
        await Deno.writeTextFile(pathname, this.makeSvgString());
    }
}